src/main/java/info/tehnut/soulshards/api/BindingEvent.java

```diff
@@ -1,15 +1,54 @@
 package info.tehnut.soulshards.api;
 
-import net.fabricmc.fabric.util.HandlerArray;
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
 import net.minecraft.entity.LivingEntity;
+import net.minecraft.util.ActionResult;
 import net.minecraft.util.Identifier;
 import net.minecraft.util.TypedActionResult;
 
 public class BindingEvent {
 
-    public static final HandlerArray<NewBinding> NEW_BINDING = new HandlerArray<>(NewBinding.class);
-    public static final HandlerArray<GainSouls> GAIN_SOULS = new HandlerArray<>(GainSouls.class);
-    public static final HandlerArray<GetEntityName> GET_ENTITY_NAME = new HandlerArray<>(GetEntityName.class);
+    public static final Event<NewBinding> NEW_BINDINGS = EventFactory.createArrayBacked(NewBinding.class,
+            (listeners) -> (entity, binding) -> {
+                TypedActionResult<IBinding> result = new TypedActionResult<>(ActionResult.PASS, binding);
+
+                for(NewBinding newBinding : listeners) {
+                    TypedActionResult<IBinding> currentResult = newBinding.onNewBinding(entity, binding);
+
+                    if(currentResult.getResult() != ActionResult.PASS) {
+                        result = currentResult;
+                    }
+                }
+
+                return result;
```

Comment:
This should probably return early. When handlers return SUCCESS or FAILURE, they generally expect no other code to be run. If that's not the case, the behaviour should be documented.

Reply:
Does the commit "fixing pyrofab's "return early" comment" solve this? Not fully understanding what I need to change here

Reply:
seems good to me.